### PR TITLE
Implement ability to disable notification listener via settings

### PR DESCRIPTION
**Pull request by vmaxim**

### CODE REVIEW

File: app/src/main/java/io/appium/uiautomator2/handler/UpdateSettings.java [hunk 1]
```diff
@@ -1,40 +1,50 @@
...
-            }
-
-            if (settings.containsKey(CompressedLayoutHierarchy.SETTING_NAME)) {
-                CompressedLayoutHierarchy.updateSetting((boolean) settings.get(CompressedLayoutHierarchy.SETTING_NAME));
-            }
-
-            if (settings.containsKey(WaitForIdleTimeout.SETTING_NAME)) {
-                WaitForIdleTimeout.updateSetting((int) settings.get(WaitForIdleTimeout.SETTING_NAME));
+            for (Entry<String, Object> entry : settings.entrySet()) {
+                ISetting setting = getSetting(entry.getKey());
```
Comment (mykola-mokhnach) [inline app/src/main/java/io/appium/uiautomator2/handler/UpdateSettings.java]: I'd rather assign getKey and getValue to variables here for better readability

File: app/src/main/java/io/appium/uiautomator2/handler/UpdateSettings.java [hunk 2]
```diff
@@ -44,4 +54,7 @@ public AppiumResponse safeHandle(IHttpRequest request) {
         return new AppiumResponse(getSessionId(request), WDStatus.SUCCESS, true);
     }
 
+    public ISetting getSetting(String settingName) throws IllegalAccessException, InstantiationException {
```
Comment (mykola-mokhnach) [inline app/src/main/java/io/appium/uiautomator2/handler/UpdateSettings.java]: mark as Nullable

File: app/src/main/java/io/appium/uiautomator2/model/NotificationListener.java [hunk 3]
```diff
@@ -29,11 +31,23 @@ public static NotificationListener getInstance(){
...
     }
 
     public void stop(){
+        Logger.debug("Stopping toast notification listener.");
+        if (!listener.isAlive()) {
+            Logger.debug("Toast notification listener is already stopped.");
+        }
         stopLooping = true;
+        try {
+            listener.join();
```
Comment (mykola-mokhnach) [inline app/src/main/java/io/appium/uiautomator2/model/NotificationListener.java]: is join really necessary? May it deadlock the current thread?

File: app/src/main/java/io/appium/uiautomator2/model/NotificationListener.java [hunk 4]
```diff
@@ -29,11 +31,23 @@ public static NotificationListener getInstance(){
...
+        if (listener.isAlive()) {
+            Logger.debug("Toast notification listener is already started.");
+        }
         listener.start();
     }
 
     public void stop(){
+        Logger.debug("Stopping toast notification listener.");
+        if (!listener.isAlive()) {
+            Logger.debug("Toast notification listener is already stopped.");
```
Comment (mykola-mokhnach) [inline app/src/main/java/io/appium/uiautomator2/model/NotificationListener.java]: there is no point to call join is isAlive is false

File: app/src/main/java/io/appium/uiautomator2/model/settings/AbstractSetting.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
+package io.appium.uiautomator2.model.settings;
+
+import io.appium.uiautomator2.common.exceptions.UiAutomator2Exception;
+import io.appium.uiautomator2.utils.Logger;
+
+public abstract class AbstractSetting<T> implements ISetting {
+
+    private Class<T> valueType;
```
Comment (mykola-mokhnach) [inline app/src/main/java/io/appium/uiautomator2/model/settings/AbstractSetting.java]: final

File: app/src/test/java/io/appium/uiautomator2/utils/w3c/W3CActionsPreprocessingTests.java [hunk 6]
```diff
@@ -0,0 +1,181 @@
+/*
```
Comment (mykola-mokhnach) [inline app/src/test/java/io/appium/uiautomator2/utils/w3c/W3CActionsPreprocessingTests.java]: any reason to move these tests?

File: app/src/test/java/io/appium/uiautomator2/model/settings/WaitForIdleTimeoutTests.java [hunk 7]
```diff
@@ -0,0 +1,51 @@
+package io.appium.uiautomator2.model.settings;
```
Comment (mykola-mokhnach) [inline app/src/test/java/io/appium/uiautomator2/model/settings/WaitForIdleTimeoutTests.java]: please add license header to each newly created file

File: app/src/test/java/io/appium/uiautomator2/utils/w3c/W3CActionsPreprocessingTests.java (hunk 6 above)
Comment (vmaxim) [reply app/src/test/java/io/appium/uiautomator2/utils/w3c/W3CActionsPreprocessingTests.java]: Unit tests should be located at ``src/test/java``` folder

File: app/src/main/java/io/appium/uiautomator2/handler/UpdateSettings.java [hunk 8]
```diff
@@ -1,40 +1,54 @@
...
+import io.appium.uiautomator2.model.settings.ISetting;
+import io.appium.uiautomator2.model.settings.WaitForIdleTimeout;
 import io.appium.uiautomator2.server.WDStatus;
 import io.appium.uiautomator2.utils.Logger;
-import java.util.Map;
 
 public class UpdateSettings extends SafeRequestHandler {
 
+    private static final Map<String, Class<? extends ISetting>> SETTINGS = new HashMap<String, Class<? extends ISetting>>() {
+        {
```
Comment (imurchie) [inline app/src/main/java/io/appium/uiautomator2/handler/UpdateSettings.java]: I this inner block necessary?

Comment (vmaxim) [reply app/src/main/java/io/appium/uiautomator2/handler/UpdateSettings.java]: Yes - this is  [instance initializer block](https://docs.oracle.com/javase/tutorial/java/javaOO/initial.html)

Comment (imurchie) [reply app/src/main/java/io/appium/uiautomator2/handler/UpdateSettings.java]: Ok. Thanks.

**Comment by imurchie** [comment]
What is the status of this PR?

**Comment by vmaxim** [comment]
I am done with it. @mykola-mokhnach Can you take a look please?

File: app/src/main/java/io/appium/uiautomator2/handler/UpdateSettings.java [hunk 9]
```diff
@@ -1,40 +1,54 @@
...
-                WaitForIdleTimeout.updateSetting((int) settings.get(WaitForIdleTimeout.SETTING_NAME));
+            for (Entry<String, Object> entry : settings.entrySet()) {
+                String settingName = entry.getKey();
+                Object settingValue = entry.getValue();
+                ISetting setting = getSetting(settingName);
+                if (setting != null) {
+                    setting.updateSetting(settingValue);
+                    Session.capabilities.put(settingName, settingValue);
+                } else {
+                    Logger.error(String.format("Setting '%s' is not supported.", entry.getKey()));
```
Comment (mykola-mokhnach) [inline app/src/main/java/io/appium/uiautomator2/handler/UpdateSettings.java]: shall we throw an exception instead?

File: app/src/main/java/io/appium/uiautomator2/model/NotificationListener.java [hunk 10]
```diff
@@ -29,11 +31,26 @@ public static NotificationListener getInstance(){
...
     public void stop(){
-        stopLooping = true;
+        Logger.debug("Stopping toast notification listener.");
+        if (listener == null || !listener.isAlive()) {
+            Logger.debug("Toast notification listener is already stopped.");
+            return;
+        }
+        listener.stopLooping();
+        try {
+            listener.join(WAIT_FOR_EVENT_TIMEOUT);
```
Comment (mykola-mokhnach) [inline app/src/main/java/io/appium/uiautomator2/model/NotificationListener.java]: 👍

File: app/src/main/java/io/appium/uiautomator2/model/settings/CompressedLayoutHierarchy.java [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+        super(Boolean.class);
+    }
+
+    @Override
+    protected void apply(Boolean compressLayout) {
+        // setCompressedLayoutHeirarchy doesn't exist on API <= 17
+        if (API_18) {
+            Device.getUiDevice().setCompressedLayoutHeirarchy(compressLayout);
+        } else {
+            Logger.info("SetCompressedLayoutHeirarchy doesn't exist on API <= 17");
```
Comment (mykola-mokhnach) [inline app/src/main/java/io/appium/uiautomator2/model/settings/CompressedLayoutHierarchy.java]: I'd rather prefer something like

if (!API_18) {
  throw ...
}
....

File: app/src/main/java/io/appium/uiautomator2/handler/UpdateSettings.java (hunk 9 above)
Comment (vmaxim) [reply app/src/main/java/io/appium/uiautomator2/handler/UpdateSettings.java]: From my point of view we can just skip unsupported setting like unsupported desired capabilities.

Comment (mykola-mokhnach) [reply app/src/main/java/io/appium/uiautomator2/handler/UpdateSettings.java]: unsupported caps are visible in the server log (and even there not so many people notice this). How do you think - how many of them are going to check device logs?

File: app/src/main/java/io/appium/uiautomator2/model/settings/CompressedLayoutHierarchy.java (hunk 11 above)
Comment (vmaxim) [reply app/src/main/java/io/appium/uiautomator2/model/settings/CompressedLayoutHierarchy.java]: I prefer do not change current behaviour with `throw` statement, but can remove `else` :)

File: app/src/main/java/io/appium/uiautomator2/handler/UpdateSettings.java (hunk 9 above)
Comment (vmaxim) [reply app/src/main/java/io/appium/uiautomator2/handler/UpdateSettings.java]: ok

Comment (imurchie) [reply app/src/main/java/io/appium/uiautomator2/handler/UpdateSettings.java]: This all goes through the driver, so validation/logging can happen there. Or the calling code could catch an error and log it, while continuing to function normally.

File: app/src/main/java/io/appium/uiautomator2/model/settings/CompressedLayoutHierarchy.java (hunk 11 above)
Comment (vmaxim) [reply app/src/main/java/io/appium/uiautomator2/model/settings/CompressedLayoutHierarchy.java]: Oh... `API_18` can be completely removed cuz  
https://github.com/appium/appium-uiautomator2-server/blob/master/app/build.gradle#L22
I will do this in separate PR.

**Comment by KazuCocoa** [comment]
After merging, is it good to add `enableNotificationListener` in 
 https://github.com/appium/appium/blob/master/docs/en/advanced-concepts/settings.md#supported-settings, btw?

**Pull request merged**